llvm/lib/Target/EVM/EVMStackShuffler.h

```diff
@@ -0,0 +1,540 @@
+/// of EVM opcodes this is supposed to be a `SWAP<depth>`.
+/// \p PushOrDup is a function with signature void(StackSlot const&) that is
+/// called to push or dup the slot given as its argument to the stack top.
+/// \p Pop is a function with signature void() that is called when the top most
+/// slot is popped.
+template <typename SwapT, typename PushOrDupT, typename PopT>
+void createStackLayout(Stack &CurrentStack, Stack const &TargetStack,
+                       SwapT Swap, PushOrDupT PushOrDup, PopT Pop) {
+  struct ShuffleOperations {
+    Stack &currentStack;
+    Stack const &targetStack;
+    SwapT swapCallback;
+    PushOrDupT pushOrDupCallback;
+    PopT popCallback;
+    Multiplicity multiplicity;
+
+    ShuffleOperations(Stack &CurrentStack, Stack const &TargetStack, SwapT Swap,
+                      PushOrDupT PushOrDup, PopT Pop)
+        : currentStack(CurrentStack), targetStack(TargetStack),
+          swapCallback(Swap), pushOrDupCallback(PushOrDup), popCallback(Pop) {
+      for (auto const &slot : currentStack)
+        --multiplicity[slot];
+
+      for (unsigned Offset = 0; Offset < targetStack.size(); ++Offset) {
+        auto *Slot = targetStack[Offset];
+        if (isa<JunkSlot>(Slot) && Offset < currentStack.size())
+          ++multiplicity[currentStack[Offset]];
+        else
+          ++multiplicity[Slot];
+      }
+    }
+
+    bool isCompatible(size_t Source, size_t Target) {
+      return Source < currentStack.size() && Target < targetStack.size() &&
+             (isa<JunkSlot>(targetStack[Target]) ||
+              currentStack[Source] == targetStack[Target]);
+    }
+
+    bool sourceIsSame(size_t Lhs, size_t Rhs) {
+      return currentStack[Lhs] == currentStack[Rhs];
+    }
+
+    int sourceMultiplicity(size_t Offset) {
+      return multiplicity.at(currentStack[Offset]);
+    }
+
+    int targetMultiplicity(size_t Offset) {
+      return multiplicity.at(targetStack[Offset]);
+    }
+
+    bool targetIsArbitrary(size_t Offset) {
+      return Offset < targetStack.size() && isa<JunkSlot>(targetStack[Offset]);
+    }
+
+    void swap(size_t I) {
+      swapCallback(static_cast<unsigned>(I));
+      std::swap(currentStack[currentStack.size() - I - 1], currentStack.back());
+    }
+
+    size_t sourceSize() { return currentStack.size(); }
+
+    size_t targetSize() { return targetStack.size(); }
+
+    void pop() {
+      popCallback();
+      currentStack.pop_back();
+    }
+
+    void pushOrDupTarget(size_t Offset) {
+      auto *targetSlot = targetStack[Offset];
+      pushOrDupCallback(targetSlot);
+      currentStack.push_back(targetSlot);
+    }
+  };
+
+  Shuffler<ShuffleOperations>::shuffle(CurrentStack, TargetStack, Swap,
+                                       PushOrDup, Pop);
+
+  assert(CurrentStack.size() == TargetStack.size());
+  for (unsigned I = 0; I < CurrentStack.size(); ++I) {
+    auto *Current = CurrentStack[I];
```

Comment:
I think `Current` should be defined as a reference, otherwise we assign JunkSlot address to the local variable `Current` on the line 533.